panics when invalid --policy filter is used

**Describe the bug**

as reported by @username-is-already-taken2

```
cnspec scan local --policy cnspec-policies/core/mondoo-linux-security.mql.yaml 
→ loaded configuration from /home/chris/.config/mondoo/mondoo.yml using source default
→ using service account credentials
→ discover related assets for 1 asset(s)
→ resolved assets resolved-assets=1
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x63136ee]

goroutine 1 [running]:
go.mondoo.com/cnspec/policy.(*Bundle).FilterPolicies(0x0, {0xc000f541a0, 0x1, 0xc000fd1918?})
	/home/benr/actions-runner/_work/cnspec/cnspec/policy/bundle.go:222 +0x12e
go.mondoo.com/cnspec/policy/scan.(*LocalScanner).distributeJob(0xc000863320?, 0xc000fd1b58, {0x8ba2c28?, 0xc001611ec0}, {{0x0, 0x0}, {0xc00138c180, 0x32}, {0xc000582780, 0x19}, ...})
	/home/benr/actions-runner/_work/cnspec/cnspec/policy/scan/local_scanner.go:224 +0x5e8
go.mondoo.com/cnspec/policy/scan.(*LocalScanner).Run(0x8ba2bb8?, {0x8ba2c28?, 0xc001611e60?}, 0xc000fd1b58)
	/home/benr/actions-runner/_work/cnspec/cnspec/policy/scan/local_scanner.go:142 +0x15b
go.mondoo.com/cnspec/apps/cnspec/cmd.RunScan(0xc0010221b0)
	/home/benr/actions-runner/_work/cnspec/cnspec/apps/cnspec/cmd/scan.go:539 +0x35f
go.mondoo.com/cnspec/apps/cnspec/cmd.glob..func18(0x0?, {0xc0000ffb40?, 0x0?, 0x0?}, 0x0?, 0x0?)
	/home/benr/actions-runner/_work/cnspec/cnspec/apps/cnspec/cmd/scan.go:348 +0xd9
go.mondoo.com/cnquery/apps/cnquery/cmd/builder.localProviderCmd.func1(0xc001400000?, {0xc0000ffb40?, 0x2?, 0x2?})
	/home/benr/go/pkg/mod/go.mondoo.com/cnquery@v0.0.0-20221207043121-7255f8d8fc78/apps/cnquery/cmd/builder/builder.go:193 +0x2b
github.com/spf13/cobra.(*Command).execute(0xc001400000, {0xc0000ffb20, 0x2, 0x2})
	/home/benr/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:920 +0x847
github.com/spf13/cobra.(*Command).ExecuteC(0xc192f60)
	/home/benr/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/home/benr/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
go.mondoo.com/cnspec/apps/cnspec/cmd.Execute()
	/home/benr/actions-runner/_work/cnspec/cnspec/apps/cnspec/cmd/root.go:76 +0x25
main.main()
	/home/benr/actions-runner/_work/cnspec/cnspec/apps/cnspec/cnspec.go:6 +0x17
```

**To Reproduce**

Download latest cnspec and provide an invalid `--policy` filter

**Expected behavior**

- detect that the provided policy is not available and display a warning
- if the user provides an argument with `.mql.yaml` at the end, display warning that `--policy-bundle` is probably the right argument

**Screenshots or CLI Output**

n/a

**Desktop (please complete the following information):**
- OS: [e.g. macOS]
- OS Version: [e.g. 13.0]
- Browser if applicable [e.g. Chrome, Firefox]
- Browser Version: [e.g. 106]

**Additional context**

n/a

--- a/apps/cnspec/cmd/scan.go
+++ b/apps/cnspec/cmd/scan.go
@@ -517,6 +517,14 @@ func RunScan(config *scanConfig) (*policy.ReportCollection, error) {
 		opts = append(opts, scan.WithUpstream(config.UpstreamConfig.ApiEndpoint, config.UpstreamConfig.SpaceMrn), scan.WithPlugins(config.UpstreamConfig.Plugins))
 	}
 
+	// show warning to the user of the policy filter container a bundle file name
+	for i := range config.PolicyNames {
+		entry := config.PolicyNames[i]
+		if strings.HasSuffix(entry, ".mql.yaml") || strings.HasSuffix(entry, ".mql.yml") {
+			log.Warn().Msg("you are using a bundle file as policy. Do you mean `--policy-bundle " + entry + "`")
+		}
+	}
+
 	scanner := scan.NewLocalScanner(opts...)
 	ctx := cnquery.SetFeatures(context.Background(), config.Features)
 

--- a/cli/reporter/print_compact.go
+++ b/cli/reporter/print_compact.go
@@ -9,6 +9,7 @@ import (
 
 	"github.com/mitchellh/mapstructure"
 	"github.com/muesli/termenv"
+	"github.com/rs/zerolog/log"
 	"go.mondoo.com/cnquery/cli/components"
 	"go.mondoo.com/cnquery/llx"
 	"go.mondoo.com/cnquery/stringx"
@@ -35,11 +36,15 @@ type defaultReporter struct {
 
 func (r *defaultReporter) print() error {
 	// catch case where the scan was not successful and no bundle was fetched from server
-	if r.data == nil || r.data.Bundle == nil {
+	if r.data == nil {
+		log.Debug().Msg("report does not contain any data")
 		return nil
 	}
 
-	r.bundle = r.data.Bundle.ToMap()
+	// this case can happen when we have only assets with errors, eg. where no valid policy was found
+	if r.data.Bundle != nil {
+		r.bundle = r.data.Bundle.ToMap()
+	}
 
 	// sort assets by name, to make it more intuitive
 	i := 0
@@ -55,6 +60,7 @@ func (r *defaultReporter) print() error {
 		return orderedAssets[i].Name < orderedAssets[j].Name
 	})
 
+	r.out.Write([]byte(NewLineCharacter))
 	if !r.isSummary {
 		r.printAssetSections(orderedAssets)
 	}
@@ -135,8 +141,9 @@ func (r *defaultReporter) printAssetSummary(assetMrn string, asset *policy.Asset
 	if !ok {
 		// If scanning the asset has failed, there will be no report, we should first look if there's an error for that target.
 		if err, ok := r.data.Errors[assetMrn]; ok {
+			errorMessage := strings.ReplaceAll(err, "\n", NewLineCharacter)
 			r.out.Write([]byte(termenv.String(fmt.Sprintf(
-				`✕ Errors:   %s`, err,
+				`✕ Errors:   %s`, errorMessage,
 			)).Foreground(r.Colors.Error).String()))
 		} else {
 			r.out.Write([]byte(fmt.Sprintf(
@@ -147,7 +154,7 @@ func (r *defaultReporter) printAssetSummary(assetMrn string, asset *policy.Asset
 		r.out.Write([]byte(NewLineCharacter))
 		return
 	}
-	if report == nil {
+	if report == nil || r.bundle == nil {
 		// the asset didn't match any policy, so no report was generated
 		return
 	}
@@ -236,8 +243,16 @@ func (r *defaultReporter) printAssetSummary(assetMrn string, asset *policy.Asset
 }
 
 func (r *defaultReporter) printAssetSections(orderedAssets []assetMrnName) {
-	r.out.Write([]byte(NewLineCharacter + NewLineCharacter))
-	queries := r.bundle.QueryMap()
+	if len(orderedAssets) == 0 || len(r.data.Errors) == len(orderedAssets) {
+		return
+	}
+
+	r.out.Write([]byte(NewLineCharacter))
+
+	var queries map[string]*policy.Mquery
+	if r.bundle != nil {
+		queries = r.bundle.QueryMap()
+	}
 
 	for _, assetMrnName := range orderedAssets {
 		assetMrn := assetMrnName.Mrn

--- a/policy/bundle.go
+++ b/policy/bundle.go
@@ -208,10 +208,11 @@ func (p *Bundle) ToMap() *PolicyBundleMap {
 // If the list of IDs is empty this function doesn't do anything.
 // This function does not remove orphaned queries from the bundle.
 func (p *Bundle) FilterPolicies(IDs []string) {
-	if len(IDs) == 0 {
+	if p == nil || len(IDs) == 0 {
 		return
 	}
 
+	log.Debug().Msg("filter policies for asset")
 	valid := make(map[string]struct{}, len(IDs))
 	for i := range IDs {
 		valid[IDs[i]] = struct{}{}
@@ -231,15 +232,19 @@ func (p *Bundle) FilterPolicies(IDs []string) {
 			uid, _ := mrn.GetResource(cur.Mrn, MRN_RESOURCE_POLICY)
 			if _, ok := valid[uid]; ok {
 				res = append(res, cur)
+				continue
 			}
 
+			log.Debug().Str("policy", cur.Mrn).Msg("policy does not match user-provided filter")
 			// if we have a MRN we do not check the UID
 			continue
 		}
 
 		if _, ok := valid[cur.Uid]; ok {
 			res = append(res, cur)
+			continue
 		}
+		log.Debug().Str("policy", cur.Uid).Msg("policy does not match user-provided filter")
 	}
 
 	p.Policies = res

--- a/policy/scan/aggregate_reporter.go
+++ b/policy/scan/aggregate_reporter.go
@@ -2,6 +2,7 @@ package scan
 
 import (
 	"github.com/hashicorp/go-multierror"
+	"github.com/rs/zerolog/log"
 	"go.mondoo.com/cnquery/motor/asset"
 	"go.mondoo.com/cnspec/policy"
 )
@@ -25,6 +26,7 @@ func NewAggregateReporter() *AggregateReporter {
 }
 
 func (r *AggregateReporter) AddReport(asset *asset.Asset, results *AssetReport) {
+	log.Debug().Str("asset", asset.Name).Msg("add scan result to report")
 	r.assets[asset.Mrn] = &policy.Asset{
 		Mrn:  asset.Mrn,
 		Name: asset.Name,
@@ -40,6 +42,7 @@ func (r *AggregateReporter) AddReport(asset *asset.Asset, results *AssetReport) 
 }
 
 func (r *AggregateReporter) AddScanError(asset *asset.Asset, err error) {
+	log.Debug().Str("asset", asset.Name).Msg("add scan error to report")
 	r.assets[asset.Mrn] = &policy.Asset{
 		Mrn:  asset.Mrn,
 		Name: asset.Name,

--- a/policy/scan/local_scanner.go
+++ b/policy/scan/local_scanner.go
@@ -578,9 +578,9 @@ func noPolicyErr(availablePolicies []string, filter []string) error {
 		sb.WriteString("The following policies are available:\n")
 		for i := range availablePolicies {
 			policyMrn := availablePolicies[i]
-			sb.WriteString("- " + policyMrn)
+			sb.WriteString("- " + policyMrn + "\n")
 		}
-		sb.WriteString("\n\n")
+		sb.WriteString("\n")
 	} else {
 		sb.WriteString("The policy bundle for the asset does not contain any policies\n\n")
 	}
@@ -589,7 +589,7 @@ func noPolicyErr(availablePolicies []string, filter []string) error {
 		sb.WriteString("User selected policies that are allowed to run:\n")
 		for i := range filter {
 			policyMrn := filter[i]
-			sb.WriteString("- " + policyMrn)
+			sb.WriteString("- " + policyMrn + "\n")
 		}
 		sb.WriteString("\n")
 	}

--- a/policy/scan/local_scanner.go
+++ b/policy/scan/local_scanner.go
@@ -221,7 +221,6 @@ func (s *LocalScanner) distributeJob(job *Job, ctx context.Context, upstreamConf
 	default:
 		return nil, false, errors.Errorf("unknown report type: %s", job.ReportType)
 	}
-	job.Bundle.FilterPolicies(job.PolicyFilters)
 
 	for i := range assetList {
 		asset := assetList[i]
@@ -247,6 +246,7 @@ func (s *LocalScanner) distributeJob(job *Job, ctx context.Context, upstreamConf
 		})
 	}
 
+	log.Debug().Msg("completed scanning all assets")
 	return reporter.Reports(), true, nil
 }
 
@@ -327,7 +327,7 @@ func (s *LocalScanner) RunAssetJob(job *AssetJob) {
 			job.connection = m
 			results, err := s.runMotorizedAsset(job)
 			if err != nil {
-				log.Warn().Err(err).Str("asset", job.Asset.Name).Msg("could not scan asset")
+				log.Error().Str("asset", job.Asset.Name).Msg("could not complete scan for asset")
 				job.Reporter.AddScanError(job.Asset, err)
 				return
 			}
@@ -379,6 +379,7 @@ func (s *LocalScanner) runMotorizedAsset(job *AssetJob) (*AssetReport, error) {
 			Runtime:  runtime,
 			Progress: progressListener,
 		}
+		log.Debug().Str("asset", job.Asset.Name).Msg("run scan")
 		res, policyErr = scanner.run()
 		return policyErr
 	})
@@ -568,6 +569,34 @@ func (s *localAssetScanner) run() (*AssetReport, error) {
 	return ar, nil
 }
 
+func noPolicyErr(availablePolicies []string, filter []string) error {
+	var sb strings.Builder
+	sb.WriteString("bundle doesn't contain any policies\n")
+	sb.WriteString("\n")
+
+	if len(availablePolicies) > 0 {
+		sb.WriteString("The following policies are available:\n")
+		for i := range availablePolicies {
+			policyMrn := availablePolicies[i]
+			sb.WriteString("- " + policyMrn)
+		}
+		sb.WriteString("\n\n")
+	} else {
+		sb.WriteString("The policy bundle for the asset does not contain any policies\n\n")
+	}
+
+	if len(filter) > 0 {
+		sb.WriteString("User selected policies that are allowed to run:\n")
+		for i := range filter {
+			policyMrn := filter[i]
+			sb.WriteString("- " + policyMrn)
+		}
+		sb.WriteString("\n")
+	}
+
+	return errors.New(sb.String())
+}
+
 func (s *localAssetScanner) prepareAsset() error {
 	var hub policy.PolicyHub = s.services
 
@@ -584,8 +613,14 @@ func (s *localAssetScanner) prepareAsset() error {
 		return errors.New("no bundle provided to run")
 	}
 
+	availablePolicies := s.job.Bundle.PolicyMRNs()
+
+	// filter bundle by user-provided policy filter
+	s.job.Bundle.FilterPolicies(s.job.PolicyFilters)
+
+	// if no policies are left, return an error
 	if len(s.job.Bundle.Policies) == 0 {
-		return errors.New("bundle doesn't contain any policies")
+		return noPolicyErr(availablePolicies, s.job.PolicyFilters)
 	}
 
 	// FIXME: we do not currently respect policy filters!

